### DISCUSSION TITLE
Performance and high availability

### DISCUSSION THREAD

**Discussion by kayrus** [emersion/maddy#375] [Q&A] [answered]
I wonder what is the performance of maddy? Let's say I need to send 2 million emails per day, can maddy handle this use case?
Is there a possibility for a horizontal scaling? What would be the bottleneck? PostgreSQL?

What if the email size is up to 150 Mb?

Can maddy use multiple SMTP relay backends and have a mapping of local-user credentials to an SMTP backend credentials? Can this mapping be provided by API or a helper?

I assume that these questions can be moved to FAQ, somewhere in documentation.

**Comment by foxcpp** [comment, marked as answer]
maddy is primarily targeting small-scale (e.g. personal email server, small organizations) deployments. I do have plans to make larger deployments feasible but so far this is not the priority.

>Let's say I need to send 2 million emails per day, can maddy handle this use case?

Assuming uniform load distribution, this means 46-47 messages per second. For outbound delivery, I expect the bottleneck to be somewhere in the local queue management but I have no data to back this up.

>What if the email size is up to 150 Mb?

I do not expect large messages to cause any specific performance issues.

>Can maddy use multiple SMTP relay backends ...

target.smtp (SMTP proxy) module supports basic fail-over with multiple servers but no load balancing so far. It should not be hard to implement a simple LB policy (e.g. random-based, roundrobin or hash-based). If you can open an "feature request" with more details on what you want to see - I can consider that for 0.5.

> have a mapping of local-user credentials to an SMTP backend credentials? Can this mapping be provided by API or a helper?

See above.
